pkg/innerring/blocktimer.go

```diff
@@ -30,31 +30,82 @@ type (
 		cnrWrapper *container.Client // to invoke stop container estimation
 		epoch      epochState        // to specify which epoch to stop, and epoch duration
 
-		stopEstimationDMul uint32 // X: X/Y of epoch in blocks
-		stopEstimationDDiv uint32 // Y: X/Y of epoch in blocks
+		stopEstimationDMul uint32 // X: X/Y of epoch in seconds
+		stopEstimationDDiv uint32 // Y: X/Y of epoch in seconds
 
 		basicIncome subEpochEventHandler
 	}
 )
 
-func newEpochTimer(args *epochTimerArgs) *timer.BlockTimer {
-	epochTimer := timer.NewBlockTimer(
-		func() (uint32, error) {
-			return uint32(args.epoch.EpochDuration()), nil
-		},
-		func() {
-			for _, handler := range args.newEpochHandlers {
-				handler()
-			}
-		},
-	)
+type deltaHandler struct {
+	m          sync.Mutex
```

Comment:
I think locking can be simplified here, you can have a queue of handlers with the only thread managing it (the same that receives block events).

Reply:
ok, as long as it is done [block by block](https://github.com/nspcc-dev/neofs-node/blob/fce19b776491b3489a79fc74dd5505ebe75add0b/pkg/morph/event/listener.go#L197-L215), it can be dropped. but the main internal mutex still must be kept, since netmap handling and ws reconnection are separate routines, and these changes may appear async at any time